src/workerd/api/sockets.c++

```diff
@@ -124,6 +164,30 @@ jsg::Promise<void> Socket::close(jsg::Lock& js) {
   }, [this](jsg::Lock& js, jsg::Value err) { return errorHandler(js, kj::mv(err)); });
 }
 
+jsg::Ref<Socket> Socket::startTls(jsg::Lock& js, jsg::Optional<TlsOptions> tlsOptions) {
+  JSG_REQUIRE(!isSecureSocket, TypeError, "Cannot startTls on a TLS socket.");
+
+  // The current socket's writable buffers need to be flushed. The socket's WritableStream is backed
+  // by an AsyncIoStream which doesn't implement any buffering, so we don't need to worry about
+  // flushing. But this is something to keep in mind in case this assumption no longer holds in
+  // the future.
+  //
+  // Detach the AsyncIoStream from the Writable/Readable streams and make them unusable.
+  writable->removeSink(js);
+  readable->detach(js);
+  closeFulfiller.resolver.resolve();
+
+  auto acceptedHostname = kj::heapString(domain);
+  KJ_IF_MAYBE(s, tlsOptions) {
+    KJ_IF_MAYBE(expectedHost, s->expectedServerHostname) {
+      acceptedHostname = kj::heapString(*expectedHost);
+    }
+  }
+  // All non-secure sockets should have a tlsStarter.
+  kj::Own<kj::AsyncIoStream> secure = KJ_ASSERT_NONNULL(tlsStarter)(acceptedHostname);
+  return setupSocket(js, kj::mv(secure), kj::mv(options), nullptr, true, kj::mv(domain));
```

Comment:
The `domain` member is guaranteed not to be needed?

Reply:
yeah, it is. Though I should make this more explicit, need another JSG_REQUIRE above to make sure this fails with a nicer error when `startTls` is called twice.